### ISSUE TITLE
Strange Performance issues with PL + FFCV 

### ISSUE THREAD

**Issue by lebrice** [Lightning-AI/lightning#14189]
## 🐛 Bug

Hey there. I'm having some very weird performance issues when trying to use PL in combination with FFCV.

Here are some of the weird results I'm getting:

| Pure for loops (200 batches) | time           |
| ---------------------------- | -------------- |
| PyTorch                      | 0:00:59.851324 |
| FFCV                         | 0:00:17.056751 |

| Training on 50 batches     | time           |
| -------------------------- | -------------- |
| Manual loop (FFCV)         | 0:00:47.665201 |
| Manual loop (Pytorch)      | 0:00:32.682359 |
| PL Trainer + DataLoaders           | 0:00:33.111235 |
| PL Trainer + FFCV                  | **0:03:05.346748** |
| PL Trainer + Obfuscated DataLoader | 0:00:31.623927 |

As you can see, there is a very weird slowdown that only happens when using FFCV in conjunction with PL. However, iterating over the ffcv Loader is faster than the PyTorch loaders (as is expected), and the manual training loop, while a bit slower, is still in a comparable range.

Here is the source code:
- ImagenetFfcvDataModule: https://gist.github.com/lebrice/37d89c29388d7fc9ce267eed1ba6dbda
- The benchmarking script: https://gist.github.com/lebrice/8b00ea2911d88fcc6b3f9c156f49b555

The ImagenetDataModule is basically the same as in pl_bolts, with some customized prepare_data that copies the dataset from where it is on our cluster at Mila, to the fast SLURM_TMPDIR directory. The code for it can be found here: https://gist.github.com/lebrice/4a67df47d9fca3e199d3e7686396240c

I've got several questions. Does someone with FFCV experience in the PL team (e.g. @carmocca ) have any idea what might be going on here?
- I've considered the possibility that PL just assumes that the dataloaders give back CPU tensors, or that PL is somehow performing an extra copy from GPU to CPU. This doesn't appear to be the case though, as far as I can tell.
- I then thought that perhaps PL is doing some optimizations on the DataLoader it receives, which would explain the difference in runtime. However, even if the dataloader is obfuscated, (wrapped in functions, etc), the runtime stays the same (Btw, this is really great, really shows that you can pass arbitrary iterables to PL, not just DataLoaders). 
- Perhaps the Trainer is having some effect on the FFCV side, e.g. by setting some global state, or initializing a torch.distributed context, that somehow affects the FFCV Loader implicitly?

### To Reproduce

1. Install FFCV
2. Pip install the ImagenetFfcvDataModule: `pip install git+https://gist.github.com/lebrice/37d89c29388d7fc9ce267eed1ba6dbda`
3. Run the benchmarking script above

### Expected behavior

I'd expect PyTorch-Lightning's Trainer to work well with FFCV.

### Environment

<details>
  <summary>Details</summary>
    * CUDA:
        - GPU:
                - Quadro RTX 8000
        - available:         True
        - version:           10.2
* Lightning:
        - lightning-bolts:   0.5.0
        - pytorch-lightning: 1.7.0
        - pytorch-pfn-extras: 0.5.8
        - torch:             1.12.1
        - torchmetrics:      0.7.3
        - torchvision:       0.13.1
* Packages:
        - absl-py:           1.2.0
        - aiohttp:           3.8.1
        - aiosignal:         1.2.0
        - assertpy:          1.1
        - async-timeout:     4.0.2
        - attrs:             22.1.0
        - autoflake:         1.4
        - black:             22.6.0
        - braceexpand:       0.1.7
        - cachetools:        5.2.0
        - certifi:           2022.6.15
        - cfgv:              3.3.1
        - charset-normalizer: 2.1.0
        - click:             8.1.3
        - cupy:              11.0.0
        - cycler:            0.11.0
        - distlib:           0.3.5
        - fairscale:         0.4.6
        - fastargs:          1.2.0
        - fastrlock:         0.8
        - ffcv:              0.0.3
        - filelock:          3.7.1
        - fonttools:         4.34.4
        - frozenlist:        1.3.0
        - fsspec:            2022.7.1
        - google-auth:       2.9.1
        - google-auth-oauthlib: 0.4.6
        - grpcio:            1.47.0
        - identify:          2.5.2
        - idna:              3.3
        - imgcat:            0.5.0
        - importlib-metadata: 4.12.0
        - iniconfig:         1.1.1
        - joblib:            1.1.0
        - kiwisolver:        1.4.4
        - lightning-bolts:   0.5.0
        - llvmlite:          0.38.1
        - markdown:          3.4.1
        - markupsafe:        2.1.1
        - matplotlib:        3.5.2
        - multidict:         6.0.2
        - mypy:              0.910
        - mypy-extensions:   0.4.3
        - nodeenv:           1.7.0
        - numba:             0.55.2
        - numpy:             1.22.4
        - oauthlib:          3.2.0
        - opencv-python:     4.6.0
        - packaging:         21.3
        - pandas:            1.4.3
        - pathspec:          0.9.0
        - pillow:            9.2.0
        - pip:               22.1.2
        - platformdirs:      2.5.2
        - pluggy:            1.0.0
        - pre-commit:        2.20.0
        - protobuf:          3.19.4
        - psutil:            5.9.0
        - py:                1.11.0
        - pyasn1:            0.4.8
        - pyasn1-modules:    0.2.8
        - pydeprecate:       0.3.2
        - pyflakes:          2.5.0
        - pyparsing:         3.0.9
        - pytest:            7.1.2
        - python-dateutil:   2.8.2
        - pytorch-lightning: 1.7.0
        - pytorch-pfn-extras: 0.5.8
        - pytz:              2022.1
        - pyyaml:            6.0
        - requests:          2.28.1
        - requests-oauthlib: 1.3.1
        - rsa:               4.9
        - scikit-learn:      1.1.1
        - scipy:             1.9.0
        - setuptools:        61.2.0
        - simple-parsing:    0.0.20
        - six:               1.16.0
        - sklearn:           0.0
        - tensorboard:       2.9.1
        - tensorboard-data-server: 0.6.1
        - tensorboard-plugin-wit: 1.8.1
        - terminaltables:    3.1.10
        - threadpoolctl:     3.1.0
        - toml:              0.10.2
        - tomli:             2.0.1
        - torch:             1.12.1
        - torchmetrics:      0.7.3
        - torchvision:       0.13.1
        - tqdm:              4.64.0
        - typing-extensions: 4.1.1
        - urllib3:           1.26.11
        - virtualenv:        20.16.2
        - webdataset:        0.2.5
        - werkzeug:          2.2.1
        - wheel:             0.37.1
        - yarl:              1.8.0
        - zipp:              3.8.1
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.9.13
        - version:           #189-Ubuntu SMP Wed May 18 14:13:57 UTC 2022
</details>

Any help would be greatly appreciated!

**Comment by awaelchli** [comment]
@lebrice One idea I have is you could use the profiler in PL to look at where the time gets spent when comparing the three PL implementations you listed.

**Comment by lebrice** [comment]
Hey @awaelchli , thanks for replying.

Here are both profiler outputs (with and without FFCV). 

<details>
  <summary>With FFCV</summary>

```
python main.py --datamodule imagenet_ffcv --max_epochs 1 --limit_train_batches=100 --limit_val_batches=10 --accelerator=gpu --n_workers=16 --profiler=pytorch

Epoch 0: 100%|████████████████████████████████| 110/110 [00:35<00:00,  3.08it/s, loss=7, v_num=2144009, train/accuracy=0.000, train/top_5_accuracy=0.000, train/loss=7.130]`Trainer.fit` stopped: `max_epochs=1` reached.                                                                                                                             
Epoch 0: 100%|████████████████████████████████| 110/110 [00:35<00:00,  3.08it/s, loss=7, v_num=2144009, train/accuracy=0.000, train/top_5_accuracy=0.000, train/loss=7.130]
FIT Profiler Report
Profile stats for: records
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                          ProfilerStep*         0.06%       1.383ms        99.65%        2.123s        1.061s       0.000us         0.00%      30.214ms      15.107ms             2  
[pl][profile][LightningModule]Model.validation_step_...         0.07%       1.535ms        98.91%        2.107s        1.054s       0.000us         0.00%       1.002ms     501.000us             2  
[pl][module]torchmetrics.classification.accuracy.Acc...         0.10%       2.230ms        98.50%        2.098s        1.049s       0.000us         0.00%     471.000us     235.500us             2  
                                       cudaLaunchKernel        97.92%        2.086s        97.92%        2.086s       2.905ms       0.000us         0.00%       0.000us       0.000us           718  
                                              aten::min         0.15%       3.276ms        97.86%        2.085s     130.310ms      58.000us         0.19%      58.000us       3.625us            16  
[pl][profile][Strategy]SingleDeviceStrategy.validati...         0.01%     207.000us         0.60%      12.744ms       6.372ms       0.000us         0.00%      26.133ms      13.066ms             2  
[pl][module]torchvision.models.resnet.ResNet: backbo...         0.02%     359.000us         0.57%      12.089ms       6.045ms       0.000us         0.00%      26.097ms      13.049ms             2  
                                        cudaMemcpyAsync         0.36%       7.622ms         0.36%       7.622ms      74.725us       0.000us         0.00%       0.000us       0.000us           102  
                                       aten::is_nonzero         0.00%      54.000us         0.35%       7.505ms     156.354us       0.000us         0.00%      50.000us       1.042us            48  
                                             aten::item         0.00%      48.000us         0.35%       7.451ms     155.229us       0.000us         0.00%      50.000us       1.042us            48  
                              aten::_local_scalar_dense         0.01%     237.000us         0.35%       7.403ms     154.229us      50.000us         0.17%      50.000us       1.042us            48  
[pl][module]torchmetrics.classification.accuracy.Acc...         0.08%       1.619ms         0.32%       6.890ms       3.445ms       0.000us         0.00%     512.000us     256.000us             2  
                                  cudaDeviceSynchronize         0.32%       6.791ms         0.32%       6.791ms       6.791ms       0.000us         0.00%       0.000us       0.000us             1  
                                           aten::conv2d         0.00%      71.000us         0.17%       3.653ms      91.325us       0.000us         0.00%      20.018ms     500.450us            40  
                                      aten::convolution         0.00%     102.000us         0.17%       3.582ms      89.550us       0.000us         0.00%      20.018ms     500.450us            40  
                                     aten::_convolution         0.01%     153.000us         0.16%       3.480ms      87.000us       0.000us         0.00%      20.018ms     500.450us            40  
                                aten::cudnn_convolution         0.09%       1.811ms         0.16%       3.327ms      83.175us      20.018ms        66.25%      20.018ms     500.450us            40  
[pl][module]torch.nn.modules.container.Sequential: b...         0.00%      64.000us         0.14%       3.036ms       1.518ms       0.000us         0.00%       9.691ms       4.846ms             2  
[pl][module]torch.nn.modules.container.Sequential: b...         0.00%      71.000us         0.13%       2.789ms       1.395ms       0.000us         0.00%       5.762ms       2.881ms             2  
[pl][module]torch.nn.modules.container.Sequential: b...         0.00%      60.000us         0.12%       2.536ms       1.268ms       0.000us         0.00%       3.471ms       1.736ms             2  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 2.130s
Self CUDA time total: 30.214ms

Done in 0:00:57.395263
```
</details>

<details>
  <summary>without FFCV</summary>

```
python main.py --datamodule imagenet --max_epochs 1 --limit_train_batches=100 --limit_val_batches=10 --accelerator=gpu --n_workers=16 --profiler=pytorch

Epoch 0: 100%|█████████████████████████████| 110/110 [00:14<00:00,  7.36it/s, loss=6.99, v_num=2144009, train/accuracy=0.000, train/top_5_accuracy=0.000, train/loss=7.130]`Trainer.fit` stopped: `max_epochs=1` reached.                                                                                                                             
Epoch 0: 100%|█████████████████████████████| 110/110 [00:14<00:00,  7.36it/s, loss=6.99, v_num=2144009, train/accuracy=0.000, train/top_5_accuracy=0.000, train/loss=7.130]
FIT Profiler Report
Profile stats for: records
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                          ProfilerStep*         0.07%       1.466ms        99.64%        2.142s        1.071s       0.000us         0.00%      30.348ms      15.174ms             2  
[pl][profile][LightningModule]Model.validation_step_...         0.07%       1.424ms        98.86%        2.125s        1.062s       0.000us         0.00%     993.000us     496.500us             2  
[pl][module]torchmetrics.classification.accuracy.Acc...         0.11%       2.374ms        98.43%        2.116s        1.058s       0.000us         0.00%     459.000us     229.500us             2  
                                       cudaLaunchKernel        97.86%        2.104s        97.86%        2.104s       2.930ms       0.000us         0.00%       0.000us       0.000us           718  
                                              aten::min         0.15%       3.189ms        97.79%        2.102s     131.368ms      55.000us         0.18%      55.000us       3.438us            16  
[pl][profile][Strategy]SingleDeviceStrategy.validati...         0.01%     228.000us         0.64%      13.705ms       6.853ms       0.000us         0.00%      26.293ms      13.146ms             2  
[pl][module]torchvision.models.resnet.ResNet: backbo...         0.02%     352.000us         0.61%      13.011ms       6.505ms       0.000us         0.00%      26.253ms      13.127ms             2  
[pl][module]torchmetrics.classification.accuracy.Acc...         0.08%       1.743ms         0.34%       7.302ms       3.651ms       0.000us         0.00%     517.000us     258.500us             2  
                                        cudaMemcpyAsync         0.33%       7.025ms         0.33%       7.025ms      68.873us       0.000us         0.00%       0.000us       0.000us           102  
                                       aten::is_nonzero         0.00%      54.000us         0.32%       6.870ms     143.125us       0.000us         0.00%      49.000us       1.021us            48  
                                             aten::item         0.00%      62.000us         0.32%       6.816ms     142.000us       0.000us         0.00%      49.000us       1.021us            48  
                              aten::_local_scalar_dense         0.01%     236.000us         0.31%       6.754ms     140.708us      49.000us         0.16%      49.000us       1.021us            48  
                                  cudaDeviceSynchronize         0.30%       6.402ms         0.30%       6.402ms       6.402ms       0.000us         0.00%       0.000us       0.000us             1  
                                           aten::conv2d         0.00%      58.000us         0.18%       3.891ms      97.275us       0.000us         0.00%      20.183ms     504.575us            40  
                                      aten::convolution         0.01%     110.000us         0.18%       3.833ms      95.825us       0.000us         0.00%      20.183ms     504.575us            40  
                                     aten::_convolution         0.01%     142.000us         0.17%       3.723ms      93.075us       0.000us         0.00%      20.183ms     504.575us            40  
                                aten::cudnn_convolution         0.09%       1.922ms         0.17%       3.581ms      89.525us      20.183ms        66.51%      20.183ms     504.575us            40  
[pl][module]torch.nn.modules.container.Sequential: b...         0.00%      75.000us         0.16%       3.381ms       1.690ms       0.000us         0.00%       9.796ms       4.898ms             2  
[pl][module]torch.nn.modules.container.Sequential: b...         0.00%      72.000us         0.14%       3.010ms       1.505ms       0.000us         0.00%       5.843ms       2.921ms             2  
[pl][module]torch.nn.modules.container.Sequential: b...         0.00%      68.000us         0.13%       2.729ms       1.365ms       0.000us         0.00%       3.469ms       1.734ms             2  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 2.149s
Self CUDA time total: 30.348ms

Done in 0:00:33.309257
```
</details>

I'm not super well versed in reading these outputs, but as far as I can tell, there doesn't seem to be anything drastically different between these two outputs that would explain the discrepancy.

(@carmocca , @obilaniu , @andrewilyas) Anyone have some idea of what could be going on here?

**Comment by Borda** [comment]
would be interested if we have this issue still with 2.0, @aniketmaurya? :rabbit:
